README.md
Yul+
===

![Node.js CI](https://github.com/FuelLabs/yulp/workflows/Node.js%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/yulp.svg)](https://badge.fury.io/js/yulp)

A low-level, highly efficient extension to Yul, an intermediate smart-contract language for the Ethereum Virtual Machine.

[Try it Now!](https://yulp.fuel.sh)

# Features
- All existing Yul features
- Memory structures (`mstruct`)
- Enums (`enum`)
- Constants (`const`)
- Ethereum standard ABI signature/topic generation (`sig"function ..."`, `topic"event ...`)
- Booleans (`true`, `false`)
- Safe math (over/under flow protection for addition, subtraction, multiplication)
- Injected methods (`mslice` and `require`)

# Coming Soon
- Static typing
- CLI support

# Installing

```sh
npm install yulp
```

# Building From Source

```sh
npm install
npm run build
npm test
```

# Library Usage

## Code Example

```js
const yulp = require('../index');
const source = yulp.compile(`

object "SimpleStore" {
  code {
    datacopy(0, dataoffset("Runtime"), datasize("Runtime"))
    return(0, datasize("Runtime"))
  }
  object "Runtime" {
    code {
      calldatacopy(0, 0, 36) // write calldata to memory

      mstruct StoreCalldata( // Custom addressable calldata structure
        sig: 4,
        val: 32
      )

      switch StoreCalldata.sig(0) // select signature from memory (at position 0)

      case sig"function store(uint256 val)" { // new signature method
        sstore(0, StoreCalldata.val(0)) // sstore calldata value
        log2(0, 0, topic"event Store(uint256 value)", StoreCalldata.val(0))
      }

      case sig"function get() returns (uint256)" {
        mstore(100, sload(0))
        return (100, 32)
      }
    }
  }
}

`);

console.log(yulp.print(source.results));
```

## Enums

Here we have a fully featured `enum` identifier which acts as a constant.

```js
object "contract" {
  code {
    enum Colors (
      Red,
      Blue,
      Green
    )

    log1(0, 0, Colors.Green) // logs 2 as a topic
  }
}
```

## Constants

`const` will define a `let` variable value that cannot be re-assigned.

```js
object "contract" {
  code {
    const firstVar := 0xaa
    const someOther := 1
  }
}
```

## Memory Slice

`mslice(position, length)` will return a 1-32 byte value from memory.

```js
object "contract" {
  code {
    mstore(30, 0xaaaa)

    log1(0, 0, mslice(30, 2)) // will log 0xaaaa
  }
}
```

## Booleans

`true` and `false` are added and equate to values `0x01` and `0x00`.

```js
object "contract" {
  code {
    mstore(30, true)

    log1(0, 0, mload(30)) // will log 0x01
  }
}
```

## Multi mstore sugar

`mstore` can now be used as a proxy method

```js
object "contract" {
  code {
    mstore(30, 1, 2, 3, 4)

    /*
    mstore(30, 1)
    mstore(add(30, 32), 2)
    mstore(add(30, 64), 3)
    mstore(add(30, 96), 4)
    */
  }
}
```

## Comparison Methods

`lte`, `gte`, `neq` can now be used.

```
if and(lte(1, 10), gte(5, 2)) {
  let k := neq(0, 1)
}
```

## MAX_UINT

`MAX_UINT` literal is now available (i.e. `uint(-1)`)

```
if lt(v, MAX_UINT) {
  let k := 1
}
```

## Ethereum Standard ABI Signature and Topic Generation

`sig" [ method abi ] "` will equate to a 4 byte method signature hex value

`topic" [ event abi definition ] "` will equate to the 32 byte topic hash

```js
object "contract" {
  code {
    const storeSig := sig"function store(uint256 val)"
    const eventSig := topic"event Store (uint256 indexed val)"

    log1(0, 0, storeSig) // will log 0x6057361d

    log1(0, 0, eventSig) // will log 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef
  }
}
```

## Memory Structures

Memory structures enable better handling of pre-existing in-memory structures.

`mstruct Identifier ( [ property, ... ] )`

A structure property is defined by an identifier and length specifier (i.e. `blockProducer:32`) where the identifier is `blockProducer` and the length is `32 bytes`.

In-memory array like structures are defined using a `name.length` property, followed by a `name:[array item size]` property.

Note, `mstruct` properties allow for data chunk sizes up to 32 bytes only.

## Inheritance

```js
object "Utils" {
  code {
    const Size := 1

    function someMethod() -> {}
  }
}

object "SimpleStore" is "Utils" {
  code {
    mstore(0, Size)
    someMethod()
  }
}
```

## Imports

We now support basic file system usage, we don't support local path resolution just yet.

The `fs` object is simply supplied at `compile`: `yulp.compile(source, fs)`.

```js
import "./Utils.yulp"

object "SimpleStore" is "Utils" {
  ...
}
```

## Error Reporting

A new experimental (post v0.0.7) feature is the `error"some message"` literal.

This simply utf8 encodes the message to bytes, keccak256 hashes it and returns the first 4 bytes as an error identifier.

The compiler will return an `errors` property (`{ [4 byte idenfitier]: [error message], ... }`).

```js
object "contract" {
  code {
    // example structure in memory of a BlockHeader starting at mem. position 400
    mstore(400, 0xaa) // block producer
    mstore(432, 0xbb) // previous block hash
    mstore(464, 0xcc) // block height
    mstore(496, 0x03) // length of anotherArray (i.e. 3 array items)
    mstore(554, 0xaaaabbbbcccc) // array with 3 elements, 0xaaaa, 0xbbbb, 0xcccc
    mstore(560, 0xdd) // Ethereum block number
    mstore(592, 0x01) // transaction roots array length
    mstore(652, 0xffffffff) // transaction roots, one 4 byte item 0xffffffff

    mstruct BlockHeader (
      blockProducer: 32,
      previousBlockHash: 32,
      blockHeight: 32,
      anotherArray.length: 32,
      anotherArray: [2],
      ethereumBlockNumber: 32,
      roots.length: 32,
      roots: [4]
    )

    BlockHeader.blockProducer(400) // will return 0xaa
    BlockHeader.blockProducer.size() // will return 32

    BlockHeader.blockHeight(400) // will return block height 0xcc
    BlockHeader.blockHeight.offset(400) // will return pos + length
    BlockHeader.blockHeight.position(400) // will return pos
    BlockHeader.blockHeight.size() // 32
    BlockHeader.blockHeight.index() // 2
    BlockHeader.blockHeight.keccak256(500) // keccak256 hash the block height

    BlockHeader.anotherArray(400, 2) // will return anotherArray item 0xcccc

    BlockHeader.size(400) // will return the size of the entire struct
    BlockHeader.offset(400) // will return entire offset position of the struct pos + length
    BlockHeader.keccak256(400) // keccak256 hash the entire block header
  }
}
```

# Helping Out

There is always a lot of work to do, and will have many rules to maintain. So please help out in any way that you can:

- Improve documentation.
- Chime in on any open issue or pull request.
- Open new issues about your ideas for making `yulp` better, and pull requests to show us how your idea works.
- Add new tests to *absolutely anything*.
- Create or contribute to ecosystem tools.
- Spread the word!

We communicate via [issues](https://github.com/fuellabs/yulp/issues) and [pull requests](https://github.com/fuellabs/yulp/pulls).

# Donating

Please consider donating if you think Yul+ is helpful to you or that my work is valuable. We are happy if you can help us buy a cup of coffee. ❤️

- [Gitcoin grant page](https://gitcoin.co/grants/199/fuel-labs)

Or just send us some *Dai*, *USDC* or *Ether*:

- [**0x3e947a271a37Ae7B59921c57be0a3246Ee0d887C**](https://etherscan.io/address/0x3e947a271a37Ae7B59921c57be0a3246Ee0d887C)

# Coming Soon

```js
mstruct BasicRecursiveStructures ( // better structure description
  block: BlockHeader,
  root: RootHeader,
  proof: MerkleProof,
  leaf: TransactionLeaf,
  token: 32
)

mstruct SwitchStatements ( // special switch case
  data.switch: 1, // switch 0, 1, 2 for options below
  data: [InputUTXO,InputHTLC,TransactionDeposit]
)

mstruct FixedLengthArrays ( // special switch case
  someArr: (7)[32],
)
```


CHANGE_LOG.md
## 0.0.1 -- Yul Plus Launch!

1. Basic testing
2. Basic docs
3. License

## 0.0.2 -- Bug Fix

1. mstruct keccak256 bug fix
2. mstruct requireing methods fix

## 0.0.3 -- Safe Math and Require Injection Fix

1. Only injects methods into code blocks, not blocks

## 0.0.4 -- Fixed array and item hashing in mstruct

1. Fixed array and item hashing in mstruct
2. Added multi mstore
3. mstruct / mslice no longer uses safe math internally
4. testing updated

## 0.0.5 -- Code keyword

1. fixed code keyword usage (i.e. codecopy, calldata)

## 0.0.6 -- Comparison methods and MAX_UINT

1. Comparison methods: lte, gte, neq
2. MAX_UINT (i.e. `uint(-1)`)
3. Compiler now returns signatures and topics used in the contract

  `compile(...).signatures` [ { abi: .., signature: .. }, .. ]
  `compile(...).topics` [ { abi: ..., topic: ... }, .. ]

It will simply return all signatures and topics detected.

Scoping with object names coming soon.
4. signature / topic signature fixes

## 0.0.7 Better Yul support, bugs and truly global constants
1) Better support for no-identifier blocks (i.e. ` code { { .. } { .. } }`)
2) Truly global `const` replacements when the value is a literal, similar to Enum
3) Experimental `error""` literal for error reporting. Can be used to define long string error messages that get compiled down to a 4 byte identifier.

Good for cases like

```
mstore(0, error"this is a really long utf8 encoded error message")
revert(0, 32)
```

Compiler results now have an `.errors` property, with an object `{ [4 byte error hash]: 'error code message' }`.
4) If an array is under 32 bytes, you can now use the quick .slice option as follows:

mstore(0, 0x03aabbccdd)

mstruct ExampleArr (
    val.length: 1,
    val: [1]
)

ExampleArr.val.slice(0) // will return 0xaabbccdd

This is nice for byte array handling, where an mslice will can handle the value.

Note, no error checking is preformed here, the user has to check if the array length is greater than 32 bytes.

## 0.1.0 -- Inheritance

1. Basic Inheritance

```js
object "Nick" {
  code {
    const HelloWorld := 1
  }
}

object "John" is "Nick" {
  code {
    mstore(0, HelloWorld)
  }
}
```

2. Basic Imports using `node` using `fs` API.

Simply provide a `fs` object to the compile for basic imports.

```js
yulp.compile(src);

// with file imports
const fs = require('fs');
yulp.compile(src, fs);
```

3. Error code reporting

```js
require(eq(1, 1), error"one-not-one")
```

4. `mstruct` stack optimizations

## 0.1.0 -- Small Updates and Upgrades

- error hashes are now using a counting nonce (saves many bytes), same error code security
- some mstruct tuning
- updated testing


yulp.js
// Generated automatically by nearley, version 2.19.1
// http://github.com/Hardmath123/nearley
(function () {
function id(x) { return x[0]; }

  const moo = require('moo')
  const { utils } = require('ethers');
  function id(x) {return x[0]; }

  let lexer = moo.compile({
    space: { match: /\s+/, lineBreaks: true },
    singleLineComment: /\/\/.*?$/,
    multiComment: /\/\*[\s\S]*?\*\/|(?:[^\\:]|^)\/\/.*$/,
    NumberLiteral: /(?!0x)[0-9]+/,
    HexLiteral: /(?:hex)(?:"|')[0-9a-fA-F]+(?:"|')/,
    HexNumber: /0[x][0-9a-fA-F]+/,
    StringLiteral: /(?:"|').*(?:"|')/,
    SigLiteral: /(?:sig)(?:"|').*(?:"|')/,
    TopicLiteral: /(?:topic)(?:"|').*(?:"|')/,
    ":=": ":=",
    "->": "->",
    ",": ",",
    ":": ":",
    ".": ".",
    boolean: ["true", "false"],
    bracket: ["{", "}", "(", ")"],
    keyword: ["object", "code", 'let', "for", "function", "data",
      "const", "enum", "mstruct",
      "if", "else", "break", "continue", "default", "switch", "case"],
    Identifier: /(?![.])[\w.]+/,
  });

  function stringToSig(str) {
    const clean = str.trim();

    if (clean.indexOf("event") === 0) {
      const inter = new utils.Interface([str]);
      return inter.events[Object.keys(inter.events)[0]].topic;
    } else {
      const inter = new utils.Interface([str,]);
      return inter.functions[Object.keys(inter.functions)[0]].sighash;
    }
  }



function FunctionCall(d) {
  const clean = d.filter(v => v.type !== 'space' && v.type !== 'keyword');
  let property = null;

  // Appendage Property
  if (clean.slice(-1)[0].type === 'Identifier') {
    property = clean[clean.length - 1].value;
    clean[clean.length - 1] = {
      type: 'space',
      value: '',
      text: '',
    };
    clean[clean.length - 2] = {
      type: 'space',
      value: '',
      text: '',
    };
    clean[clean.length - 3] = {
      type: 'space',
      value: '',
      text: '',
    };
  }

  Object.assign(clean[0], {
    type: 'FunctionCallIdentifier',
    name: clean[0].value,
    property: property,
    value: clean[0].value,
    text:  clean[0].text,
  });

  return clean;
}

function flatDeep(arr, d = 1) {
   return d > 0 ? arr.reduce((acc, val) => acc.concat(Array.isArray(val) ? flatDeep(val, d - 1) : val), [])
                : arr.slice();
};

const allConstants = arr => flatDeep(arr, Infinity)
  .filter(v => v.type === 'constant')
  .reduce((acc, v) => Object.assign(acc, { [v.name.value]: v }), {});

const replaceOutConstants = (arr, constants) => arr
  .map(v => Array.isArray(v) ? replaceOutConstants(v, constants)
    : (v.type === 'Identifier' && constants[v.value] ? constants[v.value].value : (
      v.type === 'constant' ? '' : v
    )));

const allMemoryStructs = arr => flatDeep(arr, Infinity)
  .filter(v => v.type === 'MemoryStructure')
  .reduce((acc, v) => Object.assign(acc, { [v.name]: v }), {});

const replaceOutMemoryStructs = (arr, structs) => arr
  .map(v => Array.isArray(v) ? replaceOutMemoryStructs(v, structs)
    : (v.type === 'FunctionCallIdentifier' && structs[v.value]
      ? (v.property ? { value: v.value + '___' + v.property } : { value: v.value })
    : v));
var grammar = {
    Lexer: lexer,
    ParserRules: [
    {"name": "Yul$ebnf$1", "symbols": []},
    {"name": "Yul$ebnf$1$subexpression$1", "symbols": ["_", "Chunk"]},
    {"name": "Yul$ebnf$1", "symbols": ["Yul$ebnf$1", "Yul$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "Yul", "symbols": ["Yul$ebnf$1", "_"], "postprocess": function(d) { return d; }},
    {"name": "Chunk", "symbols": ["Comment"]},
    {"name": "Chunk", "symbols": ["ObjectDefinition"], "postprocess": function(d) { return d[0]; }},
    {"name": "ObjectDefinition$ebnf$1", "symbols": []},
    {"name": "ObjectDefinition$ebnf$1$subexpression$1", "symbols": ["_", "ObjectStatement"]},
    {"name": "ObjectDefinition$ebnf$1", "symbols": ["ObjectDefinition$ebnf$1", "ObjectDefinition$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "ObjectDefinition", "symbols": [{"literal":"object"}, "_", (lexer.has("StringLiteral") ? {type: "StringLiteral"} : StringLiteral), "_", {"literal":"{"}, "ObjectDefinition$ebnf$1", "_", {"literal":"}"}], "postprocess": function(d) { console.log(d); return d; }},
    {"name": "ObjectStatement", "symbols": ["CodeDefinition"], "postprocess": function(d) { return d[0]; }},
    {"name": "ObjectStatement", "symbols": ["DataDeclaration"], "postprocess": function(d) { return d[0]; }},
    {"name": "ObjectStatement", "symbols": ["ObjectDefinition"], "postprocess": function(d) { return d[0]; }},
    {"name": "ObjectStatement", "symbols": ["Comment"], "postprocess": function(d) { return d[0]; }},
    {"name": "DataDeclaration$subexpression$1", "symbols": ["_"]},
    {"name": "DataDeclaration$subexpression$1", "symbols": [(lexer.has("StringLiteral") ? {type: "StringLiteral"} : StringLiteral)]},
    {"name": "DataDeclaration$subexpression$2", "symbols": [(lexer.has("HexLiteral") ? {type: "HexLiteral"} : HexLiteral)]},
    {"name": "DataDeclaration$subexpression$2", "symbols": [(lexer.has("StringLiteral") ? {type: "StringLiteral"} : StringLiteral)]},
    {"name": "DataDeclaration", "symbols": [{"literal":"data"}, "_", "DataDeclaration$subexpression$1", "_", "DataDeclaration$subexpression$2"]},
    {"name": "CodeDefinition", "symbols": [{"literal":"code"}, "_", {"literal":"{"}, "_", {"literal":"}"}], "postprocess": function(d) { return null; }},
    {"name": "CodeDefinition", "symbols": [{"literal":"code"}, "_", "Block"], "postprocess": function(d) { return [d[0], d[2]]; }},
    {"name": "Comment", "symbols": ["MultiLineComment"]},
    {"name": "Comment", "symbols": ["SingleLineComment"], "postprocess": id},
    {"name": "Boolean", "symbols": [(lexer.has("boolean") ? {type: "boolean"} : boolean)], "postprocess":  function(d) {
          if (d[0].value === "true") {
            return { type: 'HexNumber', value: '0x01', text: '0x01' };
          } else {
            return { type: 'HexNumber', value: '0x00', text: '0x00' };
          }
        } },
    {"name": "SigLiteral", "symbols": [(lexer.has("SigLiteral") ? {type: "SigLiteral"} : SigLiteral)], "postprocess": 
        function(d) {
          const sig = stringToSig(d[0].value.trim().slice(4).slice(0, -1));
          return { type: 'HexNumber', value: sig, text: sig };
        }
        },
    {"name": "TopicLiteral", "symbols": [(lexer.has("TopicLiteral") ? {type: "TopicLiteral"} : TopicLiteral)], "postprocess": 
        function(d) {
          const sig = stringToSig(d[0].value.trim().slice(4).slice(0, -1));
          return { type: 'HexNumber', value: sig, text: sig };
        }
        },
    {"name": "MultiLineComment", "symbols": [(lexer.has("multiComment") ? {type: "multiComment"} : multiComment)], "postprocess": id},
    {"name": "SingleLineComment", "symbols": [(lexer.has("singleLineComment") ? {type: "singleLineComment"} : singleLineComment)], "postprocess": id},
    {"name": "FunctionCall$ebnf$1", "symbols": []},
    {"name": "FunctionCall$ebnf$1$subexpression$1", "symbols": ["_", {"literal":","}, "_", "Expression"]},
    {"name": "FunctionCall$ebnf$1", "symbols": ["FunctionCall$ebnf$1", "FunctionCall$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "FunctionCall", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "Expression", "FunctionCall$ebnf$1", "_", {"literal":")"}], "postprocess": FunctionCall},
    {"name": "FunctionCall$ebnf$2", "symbols": []},
    {"name": "FunctionCall$ebnf$2$subexpression$1", "symbols": ["_", {"literal":","}, "_", "Expression"]},
    {"name": "FunctionCall$ebnf$2", "symbols": ["FunctionCall$ebnf$2", "FunctionCall$ebnf$2$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "FunctionCall", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "Expression", "FunctionCall$ebnf$2", "_", {"literal":")"}, {"literal":"."}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier)], "postprocess": FunctionCall},
    {"name": "FunctionCall", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", {"literal":")"}], "postprocess": FunctionCall},
    {"name": "FunctionCall", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", {"literal":")"}, {"literal":"."}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier)], "postprocess": FunctionCall},
    {"name": "Assignment", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":":="}, "_", "Expression"]},
    {"name": "VariableDeclaration", "symbols": [{"literal":"let"}, "_", "IdentifierList", "_", {"literal":":="}, "_", "Expression"]},
    {"name": "ConstantDeclaration", "symbols": [{"literal":"const"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":":="}, "_", "Literal"], "postprocess": 
        function(d) {
          const clean = d.filter(v => v.type !== 'space' && v.type !== 'keyword');
          return { type: 'constant', name: clean[0], value: clean[2] };
        }
        },
    {"name": "EnumDeclaration", "symbols": [{"literal":"enum"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "IdentifierList", "_", {"literal":")"}], "postprocess": 
        function(d) {
          const clean = flatDeep(d, Infinity)
            .filter(v => v.type === 'Identifier');
          const enumName = clean[0].value;
          const vals = clean.slice(1);
        
          return vals.map((enumValue, enumIndex) => ({
            type: 'constant',
            name: {
              type: 'Identifier',
              value: enumName + '.' + enumValue.value,
              text: enumName + '.' + enumValue.value,
            },
            value: {
              type: 'NumberLiteral',
              value: String(enumIndex),
              text: String(enumIndex),
            },
          }));
        }
        },
    {"name": "MemoryStructDeclaration$ebnf$1", "symbols": []},
    {"name": "MemoryStructDeclaration$ebnf$1$subexpression$1", "symbols": ["_", {"literal":","}, "_", "MemoryStructExpression"]},
    {"name": "MemoryStructDeclaration$ebnf$1", "symbols": ["MemoryStructDeclaration$ebnf$1", "MemoryStructDeclaration$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "MemoryStructDeclaration", "symbols": [{"literal":"mstruct"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "MemoryStructExpression", "MemoryStructDeclaration$ebnf$1", "_", {"literal":")"}], "postprocess": 
          function(d) {
            const clean = flatDeep(d, Infinity)
              .filter(v => v.type === 'Identifier' || v.type === 'NumberLiteral' || v.type === 'HexNumber');
            const mstructName = clean[0].value;
            const propNames = clean.slice(1).filter(v => v.type === 'Identifier');
            const propValues = clean.slice(1).filter(v => v.type === 'NumberLiteral' || v.type === 'HexNumber');
            const propValuesBN = propValues
              .map(v => utils.bigNumberify(v.value));
        
            // Check value sizes
            for (var i = 0; i < propValuesBN.length; i++) {
              if (propValuesBN[i].gt(utils.bigNumberify(32))) {
                throw new Error('Invalid memory structure declaration: "' + propValues[i].value + '" greater than 32 in ' + mstructName);
              }
            }
            const zero = utils.bigNumberify(0);
            const length = propValuesBN.reduce((a, b) => a.add(b), zero);
        
            const offsets = propValuesBN.map((v, i) => propValuesBN.slice(0, i).reduce((a, b) => a.add(b), zero))
            const props = propNames.reduce((acc, v, i) => Object.assign(acc, {
              [v]: {
                value: propValues[i],
                offset: {
                  type: 'HexNumber',
                  text: offsets[i].toHexString(),
                  value: offsets[i].toHexString(),
                }
              },
            }), {});
            let methods = propNames.map((name, i) => `
        function ${mstructName}___${name}(position) -> result {
          result := div(mload(add(position, ${offsets[i].toHexString()})), exp(2, safeSub(256, safeMul(${propValuesBN[i].toHexString()}, 8))))
        }
        `);
            methods += `
              function ${mstructName}(position) -> result {
                result := position
              }
            `;
        
            return propNames.map((structProp, propIndex) => ({
              type: 'constant',
              name: {
                type: 'Identifier',
                value: mstructName + '.' + structProp.value,
                text: mstructName + '.' + structProp.value,
              },
              value: propValues[propIndex],
            }))
            .concat([
              {
                type: 'constant',
                name: {
                  type: 'Identifier',
                  value: mstructName + '.' + 'size',
                  text: mstructName + '.' + 'size',
                },
                value: {
                  type: 'HexNumber',
                  value: length.toHexString(),
                  text: length.toHexString(),
                },
              },
              {
                type: 'MemoryStructure',
                value: methods,
                name: mstructName,
                props,
              }
            ]);
          }
        },
    {"name": "MemoryStructExpression", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":":"}, "_", "Literal"]},
    {"name": "IdentifierList$ebnf$1", "symbols": []},
    {"name": "IdentifierList$ebnf$1$subexpression$1", "symbols": ["_", {"literal":","}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier)]},
    {"name": "IdentifierList$ebnf$1", "symbols": ["IdentifierList$ebnf$1", "IdentifierList$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "IdentifierList", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "IdentifierList$ebnf$1"]},
    {"name": "ForLoop$ebnf$1", "symbols": []},
    {"name": "ForLoop$ebnf$1$subexpression$1", "symbols": ["_", "Statement"]},
    {"name": "ForLoop$ebnf$1", "symbols": ["ForLoop$ebnf$1", "ForLoop$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "ForLoop$ebnf$2", "symbols": []},
    {"name": "ForLoop$ebnf$2$subexpression$1", "symbols": ["_", "Statement"]},
    {"name": "ForLoop$ebnf$2", "symbols": ["ForLoop$ebnf$2", "ForLoop$ebnf$2$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "ForLoop", "symbols": [{"literal":"for"}, "_", {"literal":"{"}, "ForLoop$ebnf$1", "_", {"literal":"}"}, "_", "Expression", "_", {"literal":"{"}, "ForLoop$ebnf$2", "_", {"literal":"}"}, "_", "Block"]},
    {"name": "BreakContinue", "symbols": [{"literal":"break"}]},
    {"name": "BreakContinue", "symbols": [{"literal":"continue"}]},
    {"name": "Switch$ebnf$1", "symbols": []},
    {"name": "Switch$ebnf$1$subexpression$1", "symbols": ["_", "SwitchDefinition"]},
    {"name": "Switch$ebnf$1", "symbols": ["Switch$ebnf$1", "Switch$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "Switch", "symbols": [{"literal":"switch"}, "_", "Expression", "_", "Switch$ebnf$1"]},
    {"name": "Switch$ebnf$2", "symbols": []},
    {"name": "Switch$ebnf$2$subexpression$1", "symbols": ["_", "SwitchDefinition"]},
    {"name": "Switch$ebnf$2", "symbols": ["Switch$ebnf$2", "Switch$ebnf$2$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "Switch", "symbols": [{"literal":"switch"}, "_", "Expression", "_", "Comment", "_", "Switch$ebnf$2"]},
    {"name": "SwitchDefinition", "symbols": ["Case"]},
    {"name": "SwitchDefinition", "symbols": ["Default"]},
    {"name": "Case", "symbols": [{"literal":"case"}, "_", "CaseExpression", "_", "Block"]},
    {"name": "CaseExpression", "symbols": ["Literal"]},
    {"name": "CaseExpression", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier)]},
    {"name": "Default", "symbols": [{"literal":"default"}, "_", "Block"]},
    {"name": "If", "symbols": [{"literal":"if"}, "_", "Expression", "_", "Block"]},
    {"name": "If", "symbols": [{"literal":"if"}, "_", "Expression", "_", {"literal":"{"}, "_", {"literal":"}"}]},
    {"name": "Literal", "symbols": [(lexer.has("StringLiteral") ? {type: "StringLiteral"} : StringLiteral)], "postprocess": id},
    {"name": "Literal", "symbols": [(lexer.has("NumberLiteral") ? {type: "NumberLiteral"} : NumberLiteral)], "postprocess": id},
    {"name": "Literal", "symbols": [(lexer.has("HexNumber") ? {type: "HexNumber"} : HexNumber)], "postprocess": id},
    {"name": "Literal", "symbols": ["Boolean"]},
    {"name": "Literal", "symbols": ["SigLiteral"]},
    {"name": "Literal", "symbols": ["TopicLiteral"]},
    {"name": "Expression", "symbols": ["Literal"], "postprocess": id},
    {"name": "Expression", "symbols": [(lexer.has("Identifier") ? {type: "Identifier"} : Identifier)], "postprocess": id},
    {"name": "Expression", "symbols": ["FunctionCall"], "postprocess": id},
    {"name": "Expression", "symbols": ["Comment"], "postprocess": id},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "IdentifierList", "_", {"literal":")"}, "_", {"literal":"->"}, "_", {"literal":"("}, "_", "IdentifierList", "_", {"literal":")"}, "_", "Block"]},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", {"literal":")"}, "_", {"literal":"->"}, "_", {"literal":"("}, "_", "IdentifierList", "_", {"literal":")"}, "_", "Block"]},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", {"literal":")"}, "_", {"literal":"->"}, "_", {"literal":"("}, "_", {"literal":")"}, "_", "Block"]},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "IdentifierList", "_", {"literal":")"}, "_", {"literal":"->"}, "_", {"literal":"("}, "_", {"literal":")"}, "_", "Block"]},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", "IdentifierList", "_", {"literal":")"}, "_", "Block"]},
    {"name": "FunctionDefinition", "symbols": [{"literal":"function"}, "_", (lexer.has("Identifier") ? {type: "Identifier"} : Identifier), "_", {"literal":"("}, "_", {"literal":")"}, "_", "Block"]},
    {"name": "Block$ebnf$1", "symbols": []},
    {"name": "Block$ebnf$1$subexpression$1", "symbols": ["_", "Statement"]},
    {"name": "Block$ebnf$1", "symbols": ["Block$ebnf$1", "Block$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "Block", "symbols": [{"literal":"{"}, "Block$ebnf$1", "_", {"literal":"}"}], "postprocess": 
        function(d) {
          const constants = allConstants(d);
          const structs = allMemoryStructs(d);
        
          return replaceOutMemoryStructs(replaceOutConstants(d, constants), structs);
        }
        },
    {"name": "Statement", "symbols": ["Expression"]},
    {"name": "Statement", "symbols": ["Assignment"]},
    {"name": "Statement", "symbols": ["VariableDeclaration"]},
    {"name": "Statement", "symbols": ["ConstantDeclaration"]},
    {"name": "Statement", "symbols": ["FunctionDefinition"]},
    {"name": "Statement", "symbols": ["EnumDeclaration"]},
    {"name": "Statement", "symbols": ["MemoryStructDeclaration"]},
    {"name": "Statement", "symbols": ["If"]},
    {"name": "Statement", "symbols": ["ForLoop"]},
    {"name": "Statement", "symbols": ["Switch"]},
    {"name": "Statement", "symbols": ["Comment"]},
    {"name": "_", "symbols": []},
    {"name": "_", "symbols": [(lexer.has("space") ? {type: "space"} : space)], "postprocess": id}
]
  , ParserStart: "Yul"
}
if (typeof module !== 'undefined'&& typeof module.exports !== 'undefined') {
   module.exports = grammar;
} else {
   window.grammar = grammar;
}
})();
